src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php

```diff
@@ -19,6 +19,8 @@
  *
  * It provides methods to common features needed in controllers.
  *
+ * @deprecated since Symfony 4.2. Use {@see AbstractController} instead.
```

Comment:
shouldn't we add a runtime deprecation notice also?

Reply:
DebugClassLoader will warn when you extend the class already (and this class is abstract, so it can only be used by extending it)

Reply:
yes, but still: when we can, we should add the notice

Reply:
but wouldn't this report duplicate notice all the time then ?

Reply:
The issue with not sending the notice is that we cannot spot in our own code when we extend `Controller`. Do we? Only the runtime notice would tell, at least with the current strategy in place.
The maybe double notice is not an issue IMHO. Better one more than a missed notice leading to a fatal error when moving to next major.